services/user/Registry/plugin/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+};
+use bindings::transact::plugin::intf as Transact;
+use chrono::DateTime;
+use psibase::fracpack::Pack;
+use registry as RegistryService;
+use registry::service::AppMetadata as ServiceAppMetadata;
+use registry::service::AppStatusU32 as ServiceAppStatusU32;
+use registry::service::TagRecord;
+
+use serde::Deserialize;
+
+mod errors;
+use errors::ErrorType::*;
+
+struct RegistryPlugin;
+
+trait TryParseGqlResponse: Sized {
+    fn from_gql(s: String) -> Result<Self, CommonTypes::Error>;
+}
+
+#[derive(Deserialize, Debug)]
+struct ResponseRoot<T> {
+    data: T,
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Debug)]
+struct AppMetadataResponseData {
+    metadata: ServiceAppMetadata,
+    tags: Vec<TagRecord>,
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Debug)]
+struct AppMetadataResponse {
+    appMetadata: Option<AppMetadataResponseData>,
+}
+
+impl TryParseGqlResponse for AppMetadataResponseData {
+    fn from_gql(response: String) -> Result<Self, CommonTypes::Error> {
+        // println!("response: {:?}", response);
+        let response_root: ResponseRoot<AppMetadataResponse> = serde_json::from_str(&response)
+            .map_err(|e| QueryDeserializeError.err(&e.to_string()))?;
+        // println!("response_root: {:?}", response_root);
+
+        match response_root.data.appMetadata {
+            Some(app_metadata) => {
+                // println!("app_metadata: {:?}", app_metadata);
+                Ok(app_metadata)
+            }
+            None => Err(NotFound.err("App metadata not found")),
+        }
+    }
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Debug)]
+struct TagsResponseData {
+    tags: Vec<String>,
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Debug)]
+struct TagsResponse {
+    allRelatedTags: TagsResponseData,
+}
+
+impl From<ServiceAppStatusU32> for AppStatus {
+    fn from(status: ServiceAppStatusU32) -> Self {
+        match status {
+            0 => AppStatus::Draft,
+            1 => AppStatus::Published,
+            2 => AppStatus::Unpublished,
+            _ => panic!("Invalid app status"),
+        }
+    }
+}
+
+impl TryParseGqlResponse for TagsResponseData {
+    fn from_gql(response: String) -> Result<Self, CommonTypes::Error> {
+        println!("response: {:?}", response);
+        let response_root: ResponseRoot<TagsResponse> = serde_json::from_str(&response)
+            .map_err(|e| QueryDeserializeError.err(&e.to_string()))?;
+        println!("response_root: {:?}", response_root);
+
+        Ok(response_root.data.allRelatedTags)
+    }
+}
+
+impl Developer for RegistryPlugin {
+    fn set_app_metadata(metadata: AppMetadata) -> Result<(), CommonTypes::Error> {
+        let owners_vec = metadata
+            .owners
+            .iter()
+            .map(|owner| owner.parse().unwrap())
+            .collect();
+        Transact::add_action_to_transaction(
+            "setMetadata",
+            &RegistryService::action_structs::setMetadata {
```

Comment:
Another organization suggestion:
`impl From<AppMetadata> for RegistryService::action_structs::setMetadata`